cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+namespace experimental {
+// joins
+
+/**
+ * @brief  Performs an inner join on the specified columns of two
+ * tables (left, right)
+ *
+ * Inner Join returns rows from both tables as long as the values
+ * in the columns being joined on match.
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, a: {1, 2, 5}
+ *          left_on: {0}
+ *          right_on: {1}
+ *          columns_in_common: { {0, 1} }
+ * Result: { a: {1, 2}, b: {1, 2} }
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, c: {1, 2, 5}
+ *          left_on: {0}
+ *          right_on: {0}
+ *          columns_in_common: { }
+ * Result: { a: {1, 2}, b: {1, 2}, c: {1, 2} }
+ *
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) if L does not exist in `left_on` or R does not exist in `right_on`.
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) such that the location of `L` within `left_on` is not equal to
+ * location of R within `right_on`
+ * @throws cudf::logic_error if number of elements in `left_on` or `right_on`
+ * mismatch.
+ * @throws cudf::logic_error if number of columns in either `left` or `right`
+ * table is 0 or exceeds MAX_JOIN_SIZE
+ * @throws std::out_of_range if element of `left_on` or `right_on` exceed the
+ * number of columns in the left or right table.
+ *
+ * @param[in] left The left table
+ * @param[in] right The right table
+ * @param[in] left_on The column indices from `left` to join on.
+ * The column from `left` indicated by `left_on[i]` will be compared against the column
+ * from `right` indicated by `right_on[i]`.
+ * @param[in] right_on The column indices from `right` to join on.
+ * The column from `right` indicated by `right_on[i]` will be compared against the column
+ * from `left` indicated by `left_on[i]`.
+ * @param[in] columns_in_common is a vector of pairs of column indices from
```

Comment:
I see the resolved conversation above... still not comfortable with this wording.  How about:

```a vector pair of column indices from `left` and `right` that are "in common".  For "common columns", only a single output column will be produced, which is gathered from `left` columns.```

As mentioned in the earlier discussion, the indices are actually into left and right.  The most likely use case is to specify columns that are used as part of the join (thus in left_on and right_on).  However, I don't see anything in the code that *requires* that (if you specified { 0, 1 } in the second example it seems like the code would suppress the output of column c and not generate an error).  It feels like that's reasonable to me.  There's no requirement that tables be properly normalized.  One could have a column that defined uniqueness between two tables and other columns that were replicated between the columns but the join only needs to use the unique key to properly combine the columns.  You might want to suppress the duplicated columns in that case.

Reply:
>  However, I don't see anything in the code that requires that (if you specified { 0, 1 } in the second example it seems like the code would suppress the output of column c and not generate an error)

That is not the case. It will throw an error as mentioned in the documentation.
```
 * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
 * (L, R) if L does not exist in `left_on` or R does not exist in `right_on`.
 * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
 * (L, R) such that the location of `L` within `left_on` is not equal to
 * location of R within `right_on`
```
I made these changes after having a discussion with @rgsl888prabhu 

The relevant section that generates this error is in `join_call_copmute_df` [here](https://github.com/kaatish/cudf/blob/439898f2aeb848de2b148af6590112a1d509f85e/cpp/src/join/join.cu#L482).

Reply:
OK, I missed that code when I scanned.

So I still feel like I have to read that description multiple times in order to understand the proper interpretation.  The indices contained in this array are indexes into `left` and `right` that are in common.  They are (currently) required to be corresponding pairs in `left_on` and `right_on`.   But they are not really indices into `left_on` and `right_on`.  The subtle distinction of the word `from` in this text description eludes me every time I read the description.


As a secondary issue, this would preclude the following scenario:
```
TableA  a: {0, 1, 2}, d: {90, 100, 500}
TableB  b: {1, 2, 3}, a: {1, 2, 5}, d: {100, 500, 700}
left_on = { 0 }
right_on = { 1 }
columns_in_common = { {0, 1} {1, 2} }
Result: { a: {1, 2}, b: {1, 2}, d: {100, 500} }
```
Which seems like a valid use case.

Of course, there's a simple workaround for that problem (drop any columns not compared as part of the join from one of the table views), so perhaps it's unimportant to solve in the C++ code.

Reply:
> The subtle distinction of the word from in this text description eludes me every time I read the description.

I definitely agree that the wording can be improved and I'm open to suggestions. I only wanted to clarify the behavior of columns_in_common before proceeding onto that discussion.

So this is what we had before in legacy documentation before I changed it :

```c++
* @param[in] columns_in_common is a vector of pairs of column indices into `left_on` and `right_on`, ....
```

Notice that the arrays being talked about here are `left_on` and `right_on` and not `left` and `right`.
According to me if you say a variable `ind` is an index into an array `arr` then the possible values of `ind` should be from `0` to one less than the length of `arr`. My intention for changing that `into` to `from`  was to clarify that pairs in `columns_in_common` are to be selected from `left_on` and `right_on`.

Perhaps combining both the ideas might help :

```
a vector pair of column indices from `left` and `right` that are "in common". For "common columns", only a single output column will be produced, which is gathered from `left` columns. For each of these pairs (L, R), L should exist in `left_on` and R should exist in `right_on`.
```